### PR TITLE
support including location descendants in UCR filter

### PR DESCRIPTION
**Pull request by czue**
http://manage.dimagi.com/default.asp?190762

This allows you to specify a new `"include_descendants"` option in location `ChoiceProvider`s which will filter the data for all descendant locations.

Best read :blowfish: by :blowfish:. 

@esoergel @dannyroberts @sheelio cc @benrudolph

### CODE REVIEW

File: corehq/apps/locations/models.py [hunk 1]
```diff
@@ -366,6 +368,25 @@ def by_location_id(cls, location_id):
...
+        (optionally including the initial queryset in the final result)
+        """
+        # this is basically copied straight from http://stackoverflow.com/a/5740724/8207
+        if not nodes:
+            return SQLLocation.objects.none()
+        filters = []
+        for n in nodes:
+            lft, rght = n.lft, n.rght
+            if include_self:
+                lft -=1
```
Comment (dimagimon) [inline corehq/apps/locations/models.py]: E225 missing whitespace around operator

**Comment by czue** [comment]
(will look into the build fail tomorrow)

**Comment by czue** [comment]
:white_check_mark:

**Comment by dannyroberts** [comment]
Looks good—seems like a pretty clean change! Looks good to go besides a couple comments

File: corehq/apps/locations/models.py [hunk 2]
```diff
@@ -1,6 +1,8 @@
 import warnings
 from functools import partial
 from couchdbkit import ResourceNotFound
+from django.db.models import Q
```
Comment (dimagimon) [inline corehq/apps/locations/models.py]: F401 'Q' imported but unused

File: corehq/apps/locations/models.py [hunk 3]
```diff
@@ -1,6 +1,8 @@
...
+import operator
```
Comment (dimagimon) [inline corehq/apps/locations/models.py]: F401 'operator' imported but unused

**Comment by czue** [comment]
this is :white_check_mark: (last two commits were just imports and deleting a test). @dannyroberts @esoergel any other feedback or is this good to go?

**Comment by dannyroberts** [comment]
Looks good!

**Pull request merged**